datafusion/src/optimizer/filter_push_down.rs

```diff
@@ -408,10 +412,20 @@ fn optimize(plan: &LogicalPlan, mut state: State) -> Result<LogicalPlan> {
             //
             // Join clauses with `Using` constraints also take advantage of this logic to make sure
             // predicates reference the shared join columns are pushed to both sides.
+            let mut null_unsafe_predicates = vec![];
             let join_side_filters = state
                 .filters
                 .iter()
                 .filter_map(|(predicate, columns)| {
+                    if join_type == &JoinType::Left
+                        || join_type == &JoinType::Right
```

Comment:
I think it would be safer to check on it being not anti or semi join here (to avoid problems whenever we want to add another join type).

Reply:
If we add another join type, we should check if it's null safe. If it's null safe, we should allow it to push down.

Reply:
Yes, that's why I suggest to add the "safe" variants here instead, to not allow the safe variant by default if we would add another type.

Even better would be IMO to extract this to a function 
with a normal (without wildcard) pattern match. This way, when a new variant is added, it will be a compile error. 

```rust
fun is_null_safe(join_type: JoinType) -> bool {
    match join_type...
}
```

It also makes sense to me that this would be a method on `JoinType`.

Reply:
> Yes, that's why I suggest to add the "safe" variants here instead, to not allow the safe variant by default if we would add another type.
> 
> Even better would be IMO to extract this to a function with a normal (without wildcard) pattern match. This way, when a new variant is added, it will be a compile error.
> 
> ```rust
> fun is_null_safe(join_type: JoinType) -> bool {
>     match join_type...
> }
> ```

Nice, I like the way!